File: backend/app/models/bench.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class Location(BaseModel):
    """Geographic location"""
    lat: float = Field(..., description="Latitude")
    lon: float = Field(..., description="Longitude")


class BenchBase(BaseModel):
    """Base bench information"""
    id: int
    osm_id: Optional[int] = None
    name: Optional[str] = None
    location: Location
    elevation: Optional[float] = None


class BenchListItem(BenchBase):
    """Bench in list view with current status"""
    distance: float = Field(..., description="Distance from query point in meters")
    current_status: str = Field(..., description="Current sun status: sunny, shady, or unknown")
    sun_until: Optional[datetime] = Field(None, description="Time until sun changes")
    remaining_minutes: Optional[int] = Field(None, description="Minutes until sun changes")
    status_note: Optional[str] = Field(None, description="Additional status note when no change is within window")


class BenchDetail(BenchBase):
    """Detailed bench information"""
    current_status: str = Field(..., description="Current sun status: sunny, shady, or unknown")
    sun_until: Optional[datetime] = Field(None, description="Time until sun changes")
    remaining_minutes: Optional[int] = Field(None, description="Minutes until sun changes")
    status_note: Optional[str] = Field(None, description="Additional status note when no change is within window")
    created_at: Optional[datetime] = None


class BenchesResponse(BaseModel):
    """Response for benches list endpoint"""
    benches: list[BenchListItem]
    window_start: Optional[datetime] = Field(None, description="Start of available precomputed window")
    window_end: Optional[datetime] = Field(None, description="End of available precomputed window")

File: mobile/lib/models/bench.dart
class BenchLocation {
  final double lat;
  final double lon;

  BenchLocation({required this.lat, required this.lon});

  factory BenchLocation.fromJson(Map<String, dynamic> json) {
    return BenchLocation(
      lat: (json['lat'] as num).toDouble(),
      lon: (json['lon'] as num).toDouble(),
    );
  }
}

class Bench {
  final int id;
  final int? osmId;
  final String? name;
  final BenchLocation location;
  final double? elevation;

  final String currentStatus; // sunny | shady | unknown
  final DateTime? sunUntil;
  final int? remainingMinutes;
  final String? statusNote;
  final DateTime? createdAt;

  Bench({
    required this.id,
    this.osmId,
    this.name,
    required this.location,
    this.elevation,
    required this.currentStatus,
    this.sunUntil,
    this.remainingMinutes,
    this.statusNote,
    this.createdAt,
  });

  factory Bench.fromJson(Map<String, dynamic> json) {
    return Bench(
      id: json['id'] as int,
      osmId: json['osm_id'] as int?,
      name: json['name'] as String?,
      location: BenchLocation.fromJson(json['location'] as Map<String, dynamic>),
      elevation: (json['elevation'] as num?)?.toDouble(),
      currentStatus: (json['current_status'] as String?) ?? 'unknown',
      sunUntil: json['sun_until'] != null ? DateTime.parse(json['sun_until']) : null,
      remainingMinutes: json['remaining_minutes'] as int?,
      statusNote: json['status_note'] as String?,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
    );
  }

  String get displayName => name ?? 'Bench #$id';
}
